benchmarks/serving/benchmark_serving.py

```diff
@@ -0,0 +1,372 @@
+        p99_ttft_ms,
+        mean_tpot_ms,
+        median_tpot_ms,
+        p99_tpot_ms,
+    )
+
+
+async def throughput_benchmark(
+    backend: str,
+    api_url: str,
+    model_id: str,
+    tokenizer: PreTrainedTokenizerBase,
+    input_requests: List[Tuple[str, int, int]],
+    best_of: int,
+    use_beam_search: bool,
+    request_rate: float,
+):
+    if backend in ASYNC_REQUEST_FUNCS:
+        request_func = ASYNC_REQUEST_FUNCS.get(backend)
+    else:
+        raise ValueError(f"Unknown backend: {backend}")
+
+    print(f"Traffic request rate: {request_rate}")
+
+    benchmark_start_time = time.perf_counter()
+    tasks = []
+    async for request in get_request(input_requests, request_rate):
+        prompt, prompt_len, output_len = request
+        request_func_kwargs = {
+            "model": model_id,
+            "prompt": prompt,
+            "api_url": api_url,
+            "prompt_len": prompt_len,
+            "output_len": output_len,
+            "best_of": best_of,
+            "use_beam_search": use_beam_search,
+        }
+        tasks.append(asyncio.create_task(request_func(**request_func_kwargs)))
+    outputs = await asyncio.gather(*tasks)
+    benchmark_duration = time.perf_counter() - benchmark_start_time
+
+    (
+        completed,
+        total_input,
+        total_output,
+        request_throughput,
+        input_throughput,
+        output_throughput,
+        mean_ttft_ms,
+        median_ttft_ms,
+        p99_ttft_ms,
+        mean_tpot_ms,
+        median_tpot_ms,
+        p99_tpot_ms,
+    ) = calculate_metrics(
+        input_requests, outputs, benchmark_duration, tokenizer
+    )
+
+    print(f"Successful requests: {completed}")
+    print(f"Benchmark duration: {benchmark_duration:2f} s")
+    print(f"Total input tokens: {total_input}")
+    print(f"Total generated tokens: {total_output}")
+    print(f"Request throughput: {request_throughput:.2f} requests/s")
+    print(f"Input token throughput: {input_throughput:.2f} tokens/s")
+    print(f"Output token throughput: {output_throughput:.2f} tokens/s")
+    print(f"Mean TTFT: {mean_ttft_ms:.2f} ms")
+    print(f"Median TTFT: {median_ttft_ms:.2f} ms")
+    print(f"P99 TTFT: {p99_ttft_ms:.2f} ms")
+    print(f"Mean TPOT: {mean_tpot_ms:.2f} ms")
+    print(f"Median TPOT: {median_tpot_ms:.2f} ms")
+    print(f"P99 TPOT: {p99_tpot_ms:.2f} ms")
+
+    result = {}
+    result["completed"] = completed
+    result["total_input"] = total_input
+    result["total_output"] = total_output
+    result["request_throughput"] = request_throughput
+    result["input_throughput"] = input_throughput
+    result["output_throughput"] = output_throughput
+    result["duration"] = benchmark_duration
+    result["mean_ttft_ms"] = mean_ttft_ms
+    result["median_ttft_ms"] = median_ttft_ms
+    result["p99_ttft_ms"] = p99_ttft_ms
+    result["mean_tpot_ms"] = mean_tpot_ms
+    result["median_tpot_ms"] = median_tpot_ms
+    result["p99_tpot_ms"] = p99_tpot_ms
```

Comment:
Quick comment: all of these strings, including the ones in the request functions, can be added as constants as top-level of `benchmarks` to be reused by other benchmark scripts.

Reply:
Great idea!